gadget/update.go

```diff
@@ -665,6 +667,143 @@ func DiskTraitsFromDeviceAndValidate(expLayout *LaidOutVolume, dev string, opts
+	dev := ""
+	for _, vs := range laidOutVol.LaidOutStructure {
+		// here it is okay that we require there to be either a partition label
+		// or a filesystem label since we require there to be a system-boot role
+		// on this volume which by definition must have a filesystem
+		structureDevice, err := FindDeviceForStructure(&vs)
+		if err == ErrDeviceNotFound {
+			continue
+		}
+		if err != nil {
+			// TODO: should this be a fatal error?
+			return nil, err
+		}
+
+		// we found a device for this structure, get the parent disk
+		// and save that as the device for this volume
+		disk, err := disks.DiskFromPartitionDeviceNode(structureDevice)
+		if err != nil {
+			// TODO: should we keep looping instead and try again with
+			// another structure? it probably wouldn't work because we found
+			// something on disk with the same name as something from the
+			// gadget.yaml, but then we failed to make a disk from that
+			// partition which suggests something is inconsistent with the
+			// state of the disk/udev database
+			return nil, err
+		}
+
+		dev = disk.KernelDeviceNode()
+		break
+	}
+
+	if dev == "" {
+		// couldn't find a disk at all, pre-UC20 we just warn about this
+		// but let the update continue
+		if preUC20 {
+			logger.Noticef("WARNING: could not identify disk for gadget asset update of volume %s", systemBootVolume)
+			return nil, errSkipUpdateProceedRefresh
+		}
+		// fatal error on UC20+
+		return nil, fmt.Errorf("could not identify disk for gadget asset update of volume %s", systemBootVolume)
+	}
+
+	// we found the device, construct the traits with validation options
+	validateOpts := &DiskVolumeValidationOptions{
+		// allow implicit system-data on pre-uc20 only
+		AllowImplicitSystemData: preUC20,
+	}
+
+	// setup encrypted structure information to perform validation if this
+	// device used encryption
+	if !preUC20 {
+		// TODO: this needs to check if the specified partitions are ICE when
+		// we support ICE too
+
+		// TODO: is this a good enough check to know if encryption was turned
+		// on? note that we can't import the boot pkg here
+		keysPattern := filepath.Join(dirs.SnapFDEDir, "*.key")
+		files, err := filepath.Glob(keysPattern)
+		if err != nil {
+			return nil, fmt.Errorf("internal glob pattern error: %v", err)
+		}
+
+		if len(files) != 0 {
+			// then we have keys for encryption stored from install mode, so we
+			// at least should have ubuntu-data encrypted
+			validateOpts.ExpectedStructureEncryption = map[string]StructureEncryptionParameters{
+				"ubuntu-data": {Method: EncryptionLUKS},
+			}
+
+			// check if ubuntu-save exists to know whether to also assume
+			// encryption for ubuntu-save too - this should always happen since
+			// we require ubuntu-save for encryption, but there was a time
+			// during UC20 development where you could have encryption without
+			// ubuntu-save, so there is a small chance such devices could have
```

Comment:
I think such gadget would fail validation now, so we would not reach here anyway.

Reply:
do you think I should get rid of this entirely and just always assume that if encryption is turned on for a device that it has ubuntu-save ?

Reply:
if we are worried about device before we shipped UC 20, I think it's ok to ignore that case, I'm sure we have other bits of code that simply assume ubuntu-save if encrpytion is on

Reply:
ok, I will simplify this check here